lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+	_uk_store_get_static_folder(library_id, folder_name)
+
+extern struct uk_store_folder *
+_uk_store_get_dynamic_folder(unsigned int library_id, const char *name);
+
+extern struct uk_store_folder *
+_uk_store_get_static_folder(unsigned int library_id, const char *name);
+
+/**
+ * Adds a folder entry to a folder list
+ *
+ * @param folder pointer to the place where to add the new entry
+ * @param folder_entry pointer to the entry to add
+ */
+#define uk_store_add_folder_entry(folder, folder_entry)	\
+	uk_list_add(&(folder_entry)->list_head, &(folder)->entry_head)
+
+/**
+ * Removes a folder entry from a folder
+ *
+ * @param folder_entry pointer to the folder entry to delete
+ */
+#define uk_store_del_folder_entry(folder_entry)	\
+	uk_list_del(&(folder_entry)->list_head)
+
+extern struct uk_store_entry *
+uk_store_get_entry(struct uk_store_folder *folder, const char *name);
+
+/**
+ * Releases an entry (decreases the refcount and sets the reference to NULL)
+ * If the refcount is 0 (it was also released by the creator), it is removed
+ * from the list and memory can be free.
+ *
+ * @param entry pointer to the entry to release
+ */
+static inline void
+uk_store_release_entry(struct uk_store_entry **entry) {
+	struct uk_store_folder_entry *res = uk_store_get_folder_entry(entry);
+
+	ukarch_dec(&res->refcount.counter);
+
+	if (ukarch_load_n(&res->refcount.counter) <= 0)
+		uk_store_del_folder_entry(res);
+
+	*entry = NULL;
+}
+
+/**
+ * Saves a new getter function in the entry
+ *
+ * @param entry the place where to store the function
+ * @param e_type the new type of the function
+ * @param func the function
+ */
+#define uk_store_update_getter(entry, e_type, func)	\
+	do {						\
+		if (unlikely(!(entry) || !(e_type)))	\
+			break;				\
+		if ((entry)->type != (__u16)(e_type))	\
+			(entry)->set.e_type = NULL;	\
+		(entry)->type = (e_type);		\
+		(entry)->get.e_type = func;		\
+	} while (0)
+
+/**
+ * Saves a new setter function in the entry
+ *
+ * @param entry the place where to save the new function
+ * @param e_type the new type of the function
+ * @param func the function
+ */
+#define uk_store_update_setter(entry, e_type, func)	\
+	do {						\
+		if (unlikely(!(entry) || !(e_type)))	\
+			break;				\
+		if ((entry)->type != (__u16)(e_type))	\
+			(entry)->get.e_type = NULL;	\
+		(entry)->type = (e_type);		\
+		(entry)->set.e_type = func;		\
+	} while (0)
+
+/**
+ * Gets the value returned by the saved function and puts it in `out`
+ * Use only if you really know the type.
+ *
+ * @param entry the entry to use
+ * @param e_type the type of the function
+ * @param out the place where to store the result
+ */
+#define uk_store_get_value_typed(entry, e_type, out)			\
+	do {								\
+		if (unlikely(!(entry) || !(e_type)))			\
+			break;						\
+		*((__##e_type *) (out)) = (entry)->get.e_type();	\
+	} while (0)
+
+extern int uk_store_get_value(struct uk_store_entry *entry, void *out);
```

Comment:
Please read my comments about the non-typed getters/setters in the C file. I think `void *` is risky and can cause easy mis-use and programming mistakes.